File: src/optimizer/zzz_models/buff.py
"""
Buff增益模型 - 纯数据定义
"""
from dataclasses import dataclass, field
from typing import Dict, Optional
from enum import Enum

from optimizer.zzz_models.base import PropertyType
from optimizer.zzz_models.property_collection import PropertyCollection


class BuffContext(Enum):
    """Buff生效环境"""
    OUT_OF_COMBAT = 1   # 局外（面板显示）
    IN_COMBAT = 2       # 局内（战斗中）


class BuffTarget:
    """Buff生效目标"""
    def __init__(
        self,
        target_self: bool = True,      # 对自身生效
        target_enemy: bool = False,    # 对敌人生效
        target_teammate: bool = False, # 对队友生效
        target_bund: bool = False,     # 对邦布生效
    ):
        self.target_self = target_self
        self.target_enemy = target_enemy
        self.target_teammate = target_teammate
        self.target_bund = target_bund


class BuffSource(Enum):
    """Buff来源"""
    AGENT_PASSIVE = 1       # 角色被动
    AGENT_TALENT = 2        # 角色天赋
    TALENT = 11             # 天赋（简写）
    AGENT_CORE = 3          # 核心技能
    WENGINE = 4             # 音擎
    WENGINE_TALENT = 5      # 音擎天赋
    DRIVE_DISK_2PC = 6      # 驱动盘2件套
    DRIVE_DISK_4PC = 7      # 驱动盘4件套
    TEAM_MATE = 8           # 队友
    CORE_PASSIVE = 9        # 核心被动
    POTENTIAL = 10          # 潜能觉醒


@dataclass
class ConversionBuff:
    """
    转换类BUFF模型
    
    用于属性转换效果（如：每点异常精通转伤害）
    不分局内局外，始终生效
    """
    id: str                                             # 唯一标识
    name: str                                           # 增益名称
    description: str                                    # 效果描述
    source: BuffSource                                  # 来源
    
    # 转换相关
    from_property: PropertyType                          # 源属性
    to_property: PropertyType                            # 目标属性
    conversion_ratio: float                              # 转换比例
    
    # 上限（可选）
    max_value: Optional[float] = None                   # 转换后的最大值
    
    # 触发条件（描述文本）
    trigger_conditions: str = ""                         # 触发条件
    
    # 生效条件
    is_active: bool = True
    
    def __post_init__(self):
        if isinstance(self.source, int):
            self.source = BuffSource(self.source)
        if isinstance(self.from_property, str):
            self.from_property = PropertyType[self.from_property]
        if isinstance(self.to_property, str):
            self.to_property = PropertyType[self.to_property]
    
    @classmethod
    def from_dict(cls, data: dict) -> 'ConversionBuff':
        """从字典创建"""
        # 转换 max_value（如果目标属性是百分比，需要从百分数转为小数）
        max_value = data.get('max_value')
        if max_value is not None:
            to_prop = PropertyType[data['to_property']] if isinstance(data['to_property'], str) else PropertyType(data['to_property'])
            if to_prop.is_percentage:
                max_value = max_value / 100.0

        return cls(
            id=data['id'],
            name=data['name'],
            description=data['description'],
            source=BuffSource[data['source']] if isinstance(data['source'], str) else BuffSource(data['source']),
            from_property=PropertyType[data['from_property']] if isinstance(data['from_property'], str) else PropertyType(data['from_property']),
            to_property=PropertyType[data['to_property']] if isinstance(data['to_property'], str) else PropertyType(data['to_property']),
            conversion_ratio=data['conversion_ratio'],
            max_value=max_value,
            trigger_conditions=data.get('trigger_conditions', ''),
            is_active=data.get('is_active', True),
        )
    
    def to_dict(self) -> dict:
        """转为字典"""
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'source': self.source.name,
            'from_property': self.from_property.name,
            'to_property': self.to_property.name,
            'conversion_ratio': self.conversion_ratio,
            'max_value': self.max_value,
            'trigger_conditions': self.trigger_conditions,
            'is_active': self.is_active,
        }
    
    def calculate_value(self, source_value: float) -> float:
        """
        计算转换后的数值
        
        Args:
            source_value: 源属性的数值
            
        Returns:
            转换后的数值
        """
        if not self.is_active:
            return 0.0
        
        # 计算转换值
        value = source_value * self.conversion_ratio
        
        # 应用上限
        if self.max_value and value > self.max_value:
            value = self.max_value
        
        return value

    def format(self, indent: int = 0) -> str:
        """格式化输出转换BUFF信息（只输出有意义的值）

        Args:
            indent: 缩进空格数

        Returns:
            格式化字符串
        """
        lines = []
        prefix = " " * indent

        # 转换BUFF基本信息
        lines.append(f"{prefix}【转换BUFF】")
        lines.append(f"  {prefix}名称: {self.name}")
        if self.description:
            lines.append(f"  {prefix}描述: {self.description}")
        lines.append(f"  {prefix}来源: {self.source.name}")
        
        # 转换信息
        lines.append(f"{prefix}【转换】")
        lines.append(f"  {prefix}源属性: {self.from_property.cn_name}")
        lines.append(f"  {prefix}目标属性: {self.to_property.cn_name}")
        lines.append(f"  {prefix}转换比例: {self.conversion_ratio * 100:.0f}%")
        
        # 上限（根据目标属性类型显示）
        if self.max_value:
            if self.to_property.is_percentage:
                # 目标属性是百分比，显示为百分比
                lines.append(f"  {prefix}上限: {self.max_value*100:.2f}%")
            else:
                # 目标属性是固定值，显示为固定值
                lines.append(f"  {prefix}上限: {self.max_value:.1f}")
        
        # 触发条件
        if self.trigger_conditions:
            lines.append(f"  {prefix}触发条件: {self.trigger_conditions}")
        
        # 生效状态
        if not self.is_active:
            lines.append(f"  {prefix}状态: 未激活")
        
        return "\n".join(lines)


@dataclass
class Buff:
    """
    增益数据模型
    
    包含所有属性加成，分为局外和局内
    """
    id: str                                             # 唯一标识
    name: str                                           # 增益名称
    description: str                                    # 效果描述
    source: BuffSource                                  # 来源
    context: BuffContext                                # 生效环境（局外/局内）
    target: BuffTarget = field(default_factory=lambda: BuffTarget())  # 生效目标（默认仅自身）
    
    # 局外属性加成（面板显示）
    out_of_combat_stats: Dict[PropertyType, float] = field(default_factory=dict)
    
    # 局内属性加成（战斗中）
    in_combat_stats: Dict[PropertyType, float] = field(default_factory=dict)
    
    # 触发条件（描述文本，为空则无条件触发）
    trigger_conditions: str = ""
    
    # 叠加层数
    max_stacks: int = 1                                 # 最大叠加层数，默认1
    stack_mode: str = "linear"                          # 叠加模式：linear（线性叠加）或 full_only（仅满层生效）
    base_value: Optional[float] = None                  # 每层基础值（用于线性叠加）
    full_stack_bonus: Optional[float] = None           # 满层额外加成
    
    # 生效条件（简化，默认全生效）
    is_active: bool = True
    
    def __post_init__(self):
        if isinstance(self.source, int):
            self.source = BuffSource(self.source)
        if isinstance(self.context, int):
            self.context = BuffContext(self.context)
        # target 已经是 BuffTarget 对象，无需转换
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Buff':
        """从字典创建"""
        # 转换 out_of_combat_stats（百分比属性需要从百分数转为小数）
        out_of_combat_stats = {}
        if 'out_of_combat_stats' in data:
            for prop_name, value in data['out_of_combat_stats'].items():
                prop_type = PropertyType[prop_name]
                # 如果是百分比属性，从百分数形式（60.0 = 60%）转为小数形式（0.6 = 60%）
                if prop_type.is_percentage:
                    out_of_combat_stats[prop_type] = value / 100.0
                else:
                    out_of_combat_stats[prop_type] = value

        # 转换 in_combat_stats（百分比属性需要从百分数转为小数）
        in_combat_stats = {}
        if 'in_combat_stats' in data:
            for prop_name, value in data['in_combat_stats'].items():
                prop_type = PropertyType[prop_name]
                # 如果是百分比属性，从百分数形式（60.0 = 60%）转为小数形式（0.6 = 60%）
                if prop_type.is_percentage:
                    in_combat_stats[prop_type] = value / 100.0
                else:
                    in_combat_stats[prop_type] = value
        
        # 转换 target
        target_data = data.get('target', {})
        target = BuffTarget(
            target_self=target_data.get('target_self', True),
            target_enemy=target_data.get('target_enemy', False),
            target_teammate=target_data.get('target_teammate', False),
            target_bund=target_data.get('target_bund', False),
        )
        
        return cls(
            id=data['id'],
            name=data['name'],
            description=data['description'],
            source=BuffSource[data['source']] if isinstance(data['source'], str) else BuffSource(data['source']),
            context=BuffContext[data['context']] if isinstance(data['context'], str) else BuffContext(data['context']),
            target=target,
            out_of_combat_stats=out_of_combat_stats,
            in_combat_stats=in_combat_stats,
            trigger_conditions=data.get('trigger_conditions', []),
            max_stacks=data.get('max_stacks', 1),
            stack_mode=data.get('stack_mode', 'linear'),
            base_value=data.get('base_value'),
            full_stack_bonus=data.get('full_stack_bonus'),
            is_active=data.get('is_active', True),
        )
    
    def to_dict(self) -> dict:
        """转为字典"""
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'source': self.source.name,
            'context': self.context.name,
            'target': {
                'target_self': self.target.target_self,
                'target_enemy': self.target.target_enemy,
                'target_teammate': self.target.target_teammate,
                'target_bund': self.target.target_bund,
            },
            'out_of_combat_stats': {prop.name: value for prop, value in self.out_of_combat_stats.items()},
            'in_combat_stats': {prop.name: value for prop, value in self.in_combat_stats.items()},
            'trigger_conditions': self.trigger_conditions,
            'max_stacks': self.max_stacks,
            'stack_mode': self.stack_mode,
            'base_value': self.base_value,
            'full_stack_bonus': self.full_stack_bonus,
            'is_active': self.is_active,
        }
    
    def calculate_value(self, current_stacks: int = 1) -> Dict[PropertyType, float]:
        """
        计算指定层数下的buff数值
        
        Args:
            current_stacks: 当前层数 (1-max_stacks)
            
        Returns:
            属性字典 {PropertyType: value}
        
        计算规则：
        - stack_mode = "linear": value = base_value * current_stacks + (满层 ? full_stack_bonus : 0)
        - stack_mode = "full_only": value = (满层 ? base_value : 0)
        """
        if not self.is_active:
            return {}
        
        stats = {}
        
        if self.stack_mode == "linear":
            # 线性叠加模式
            for prop, base_val in self.in_combat_stats.items():
                value = base_val * current_stacks
                # 满层时额外加成
                if current_stacks == self.max_stacks and self.full_stack_bonus:
                    value += self.full_stack_bonus
                stats[prop] = value
        
        elif self.stack_mode == "full_only":
            # 仅满层生效模式
            if current_stacks == self.max_stacks:
                stats = self.in_combat_stats.copy()
        
        return stats

    def to_property_collection(self) -> PropertyCollection:
        """将 Buff 转换为 PropertyCollection
        
        Returns:
            PropertyCollection：包含局外和局内属性
        """
        if not self.is_active:
            return PropertyCollection()
        
        return PropertyCollection(
            out_of_combat=self.out_of_combat_stats.copy(),
            in_combat=self.in_combat_stats.copy(),
        )

    def format(self, indent: int = 0) -> str:
        """格式化输出Buff信息（只输出有意义的值）

        Args:
            indent: 缩进空格数

        Returns:
            格式化字符串
        """
        lines = []
        prefix = " " * indent

        # Buff 基本信息
        lines.append(f"{prefix}【Buff】")
        lines.append(f"  {prefix}名称: {self.name}")
        if self.description:
            lines.append(f"  {prefix}描述: {self.description}")
        lines.append(f"  {prefix}来源: {self.source.name}")
        lines.append(f"  {prefix}环境: {self.context.name}")
        
        # 叠加信息
        if self.max_stacks > 1:
            lines.append(f"  {prefix}叠加: 最大{self.max_stacks}层 ({self.stack_mode})")
        
        # 触发条件
        if self.trigger_conditions:
            lines.append(f"  {prefix}触发条件: {self.trigger_conditions}")
        
        # 目标
        target_list = []
        if self.target.target_self:
            target_list.append("自身")
        if self.target.target_enemy:
            target_list.append("敌人")
        if self.target.target_teammate:
            target_list.append("队友")
        if self.target.target_bund:
            target_list.append("邦布")
        if target_list:
            lines.append(f"  {prefix}目标: {', '.join(target_list)}")
        
        # 属性集合（直接使用 PropertyCollection 的格式化方法）
        if self.out_of_combat_stats or self.in_combat_stats:
            lines.append(f"{prefix}【属性】")
            prop_collection = self.to_property_collection()
            formatted = prop_collection.format(indent=indent+2)
            if formatted:
                lines.append(formatted)
        
        return "\n".join(lines)

File: web/optimizer/src/model/buff.ts
/**
 * Buff增益模型 - 战斗核心数据定义
 */

import { PropertyType, isPercentageProperty, getPropertyCnName } from './base';
import { PropertyCollection } from './property-collection';

/**
 * Buff生效环境
 */
export enum BuffContext {
  OUT_OF_COMBAT = 1, // 局外（面板显示）
  IN_COMBAT = 2,     // 局内（战斗中）
}

/**
 * Buff生效目标
 */
export class BuffTarget {
  target_self: boolean = true;
  target_enemy: boolean = false;
  target_teammate: boolean = false;
  target_bund: boolean = false;
}

/**
 * Buff来源
 */
export enum BuffSource {
  AGENT_PASSIVE = 1,
  AGENT_TALENT = 2,
  TALENT = 11,
  AGENT_CORE = 3,
  WENGINE = 4,
  WENGINE_TALENT = 5,
  DRIVE_DISK_2PC = 6,
  DRIVE_DISK_4PC = 7,
  TEAM_MATE = 8,
  CORE_PASSIVE = 9,
  POTENTIAL = 10,
}

/**
 * 转换类（封装好的）
 */
export class Conversion {
  from_property: PropertyType;
  to_property: PropertyType;
  conversion_ratio: number;
  max_value?: number;
  from_property_threshold?: number; // 源属性阈值，超过此值的部分才参与转换

  constructor(
    fromProperty: PropertyType,
    toProperty: PropertyType,
    conversionRatio: number,
    maxValue?: number,
    fromPropertyThreshold?: number
  ) {
    this.from_property = fromProperty;
    this.to_property = toProperty;
    this.conversion_ratio = conversionRatio;
    this.max_value = maxValue;
    this.from_property_threshold = fromPropertyThreshold;
  }
}

/**
 * Buff
 *
 * 核心原则：
 * - 只持有对战斗有影响的属性
 * - UI相关字段通过getter从外部获取
 */
export class Buff {
  // === 战斗核心数据 ===
  id: string = '';                 // Buff ID
  name: string = '';               // Buff名称
  description: string = '';        // Buff描述
  source: BuffSource;              // 来源
  target: BuffTarget;              // 生效目标

  // 局内属性加成（战斗中生效）
  in_combat_stats: Map<PropertyType, number>;

  // 局外属性加成
  out_of_combat_stats: Map<PropertyType, number>;

  // 转换类（可选）
  conversion?: Conversion;

  // 触发条件
  trigger_conditions: string = '';

  // 叠加机制
  max_stacks: number = 1;
  stack_mode: 'linear' | 'full_only' = 'linear';

  constructor(
    source: BuffSource,
    inCombatStats?: Map<PropertyType, number> | Record<PropertyType, number>,
    outOfCombatStats?: Map<PropertyType, number> | Record<PropertyType, number>,
    conversion?: Conversion,
    target?: BuffTarget,
    maxStacks: number = 1,
    stackMode: 'linear' | 'full_only' = 'linear',
    id: string = '',
    name: string = '',
    description: string = '',
    triggerConditions: string = ''
  ) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.source = source;
    this.target = target ?? new BuffTarget();
    this.in_combat_stats = inCombatStats instanceof Map
      ? new Map(inCombatStats)
      : new Map(Object.entries(inCombatStats || {}) as unknown as [PropertyType, number][]);
    this.out_of_combat_stats = outOfCombatStats instanceof Map
      ? new Map(outOfCombatStats)
      : new Map(Object.entries(outOfCombatStats || {}) as unknown as [PropertyType, number][]);
    this.conversion = conversion;
    this.trigger_conditions = triggerConditions;
    this.max_stacks = maxStacks;
    this.stack_mode = stackMode;
  }

  // === 辅助方法 ===

  /** 判断是否是转换类BUFF */
  isConversion(): boolean {
    return this.conversion !== undefined;
  }

  /** 计算转换值 */
  calculateConversion(sourceValue: number): number {
    if (!this.conversion) return 0;

    // 如果有阈值，只有超过阈值的部分才参与转换
    const threshold = this.conversion.from_property_threshold ?? 0;
    const effectiveValue = Math.max(0, sourceValue - threshold);
    
    let value = effectiveValue * this.conversion.conversion_ratio;
    if (this.conversion.max_value !== undefined && value > this.conversion.max_value) {
      value = this.conversion.max_value;
    }
    return value;
  }

  /**
   * 应用局内属性到属性集
   *
   * @param props 输入属性集
   * @returns 应用后的属性集（新实例）
   */
  applyInCombatStats(props: PropertyCollection): PropertyCollection {
    if (this.in_combat_stats.size === 0) {
      return props;
    }

    const result = new PropertyCollection();
    result.add(props);

    for (const [prop, value] of this.in_combat_stats.entries()) {
      const currentValue = result.in_combat.get(prop) || 0;
      result.in_combat.set(prop, currentValue + value);
    }

    return result;
  }

  /**
   * 应用转换类属性到属性集
   * 基于局内属性（已计算完基础×百分比+固定值后的值）生成转换类属性
   *
   * @param props 输入属性集（应该已经包含计算后的局内属性）
   * @returns 应用后的属性集（新实例）
   */
  applyConversionStats(props: PropertyCollection): PropertyCollection {
    if (!this.conversion) {
      return props;
    }

    const result = new PropertyCollection();
    result.add(props);

    // 获取源属性的最终值（基础×百分比+固定值后的值）
    const fromProp = this.conversion.from_property;
    
    // 对于ATK_BASE等基础属性，需要计算 base * (1 + percent) + flat
    let sourceValue = 0;
    if (fromProp === PropertyType.ATK_BASE) {
      const base = props.getInCombat(PropertyType.ATK_BASE, 0);
      const percent = props.getInCombat(PropertyType.ATK_, 0);
      const flat = props.getInCombat(PropertyType.ATK, 0);
      sourceValue = base * (1 + percent) + flat;
    } else if (fromProp === PropertyType.HP_BASE) {
      const base = props.getInCombat(PropertyType.HP_BASE, 0);
      const percent = props.getInCombat(PropertyType.HP_, 0);
      const flat = props.getInCombat(PropertyType.HP, 0);
      sourceValue = base * (1 + percent) + flat;
    } else if (fromProp === PropertyType.DEF_BASE) {
      const base = props.getInCombat(PropertyType.DEF_BASE, 0);
      const percent = props.getInCombat(PropertyType.DEF_, 0);
      const flat = props.getInCombat(PropertyType.DEF, 0);
      sourceValue = base * (1 + percent) + flat;
    } else {
      sourceValue = props.getInCombat(fromProp, 0);
    }

    // 计算转换值
    const convertedValue = this.calculateConversion(sourceValue);
    
    if (convertedValue > 0) {
      result.addConversion(this.conversion.to_property, convertedValue);
    }

    return result;
  }

  /**
   * 判断是否有局内属性
   */
  hasInCombatStats(): boolean {
    return this.in_combat_stats.size > 0;
  }

  /**
   * 转换为 PropertyCollection
   *
   * @returns 属性集合
   */
  toPropertyCollection(): PropertyCollection {
    return new PropertyCollection(
      this.out_of_combat_stats,
      this.in_combat_stats
    );
  }

  /** 计算叠层后的属性值 */
  calculateValue(currentStacks: number = 1): Map<PropertyType, number> {
    const stats = new Map<PropertyType, number>();
    
    if (this.stack_mode === 'linear') {
      for (const [prop, baseVal] of this.in_combat_stats) {
        stats.set(prop, baseVal * currentStacks);
      }
    } else if (this.stack_mode === 'full_only' && currentStacks === this.max_stacks) {
      for (const [prop, val] of this.in_combat_stats) {
        stats.set(prop, val);
      }
    }
    
    return stats;
  }

  // === 序列化 ===

  static fromDict(data: any): Buff {
    // 解析conversion
    let conversion: Conversion | undefined;
    if (data.conversion) {
      conversion = new Conversion(
        (PropertyType as any)[data.conversion.from_property],
        (PropertyType as any)[data.conversion.to_property],
        data.conversion.conversion_ratio,
        data.conversion.max_value,
        data.conversion.from_property_threshold
      );
    }

    // 解析in_combat_stats（JSON中已是小数形式，如0.35表示35%）
    const inCombatStats = new Map<PropertyType, number>();
    if (data.in_combat_stats) {
      for (const [propName, value] of Object.entries(data.in_combat_stats)) {
        const propType = (PropertyType as any)[propName];
        if (propType === undefined) {
          console.warn(`[Buff解析] 未知属性类型: ${propName}，值: ${value}，来源: ${data.id || data.name}`);
          continue;
        }
        inCombatStats.set(propType, value as number);
      }
    }

    // 解析out_of_combat_stats（JSON中已是小数形式）
    const outOfCombatStats = new Map<PropertyType, number>();
    if (data.out_of_combat_stats) {
      for (const [propName, value] of Object.entries(data.out_of_combat_stats)) {
        const propType = (PropertyType as any)[propName];
        if (propType === undefined) {
          console.warn(`[Buff解析] 未知属性类型: ${propName}，值: ${value}，来源: ${data.id || data.name}`);
          continue;
        }
        outOfCombatStats.set(propType, value as number);
      }
    }

    // 解析target
    const targetData = data.target ?? {};
    const target = new BuffTarget();
    target.target_self = targetData.target_self ?? true;
    target.target_enemy = targetData.target_enemy ?? false;
    target.target_teammate = targetData.target_teammate ?? false;
    target.target_bund = targetData.target_bund ?? false;

    return new Buff(
      typeof data.source === 'string' ? (BuffSource as any)[data.source] : data.source,
      inCombatStats,
      outOfCombatStats,
      conversion,
      target,
      data.max_stacks ?? 1,
      data.stack_mode ?? 'linear',
      data.id ?? '',
      data.name ?? '',
      data.description ?? '',
      data.trigger_conditions ?? ''
    );
  }

  /**
   * 从游戏数据创建 Buff（fromDict 的别名）
   */
  static fromBuffData(data: any): Buff {
    return Buff.fromDict(data);
  }

  toDict(): any {
    const inCombatStats: Record<string, number> = {};
    for (const [prop, value] of this.in_combat_stats.entries()) {
      inCombatStats[PropertyType[prop]] = value;
    }

    const outOfCombatStats: Record<string, number> = {};
    for (const [prop, value] of this.out_of_combat_stats.entries()) {
      outOfCombatStats[PropertyType[prop]] = value;
    }

    return {
      id: this.id,
      name: this.name,
      description: this.description,
      source: BuffSource[this.source],
      target: {
        target_self: this.target.target_self,
        target_enemy: this.target.target_enemy,
        target_teammate: this.target.target_teammate,
        target_bund: this.target.target_bund,
      },
      in_combat_stats: inCombatStats,
      out_of_combat_stats: outOfCombatStats,
      conversion: this.conversion ? {
        from_property: PropertyType[this.conversion.from_property],
        to_property: PropertyType[this.conversion.to_property],
        conversion_ratio: this.conversion.conversion_ratio,
        max_value: this.conversion.max_value,
        from_property_threshold: this.conversion.from_property_threshold,
      } : null,
      trigger_conditions: this.trigger_conditions,
      max_stacks: this.max_stacks,
      stack_mode: this.stack_mode,
    };
  }

  /**
   * 格式化输出Buff信息（只输出有意义的值）
   *
   * @param indent 缩进空格数
   * @returns 格式化字符串
   */
  format(indent: number = 0): string {
    const lines: string[] = [];
    const prefix = ' '.repeat(indent);

    // 基本信息
    if (this.name) {
      lines.push(`${prefix}${this.name}`);
    }
    if (this.description) {
      lines.push(`${prefix}  ${this.description}`);
    }

    // 属性加成
    const allStats = new Map<PropertyType, number>();
    for (const [prop, value] of this.in_combat_stats.entries()) {
      allStats.set(prop, value);
    }
    for (const [prop, value] of this.out_of_combat_stats.entries()) {
      allStats.set(prop, (allStats.get(prop) ?? 0) + value);
    }

    if (allStats.size > 0) {
      lines.push(`${prefix}  属性加成:`);
      for (const [prop, value] of allStats.entries()) {
        const cnName = getPropertyCnName(prop);
        if (isPercentageProperty(prop)) {
          lines.push(`${prefix}    ${cnName}: ${(value * 100).toFixed(2)}%`);
        } else {
          lines.push(`${prefix}    ${cnName}: ${value.toFixed(2)}`);
        }
      }
    }

    return lines.join('\n');
  }
}

/**
 * 转换类 Buff - 专门用于属性转换的 Buff
 */
export class ConversionBuff extends Buff {
  // 转换详情
  conversion: Conversion;

  constructor(
    conversion: Conversion,
    source: BuffSource = BuffSource.AGENT_PASSIVE,
    target?: BuffTarget,
    maxStacks: number = 1,
    stackMode: 'linear' | 'full_only' = 'linear',
    id: string = '',
    name: string = '',
    description: string = '',
    triggerConditions: string = ''
  ) {
    super(
      source,
      undefined, // 转换Buff通常不直接提供属性加成
      undefined,
      conversion,
      target,
      maxStacks,
      stackMode,
      id,
      name,
      description,
      triggerConditions
    );
    this.conversion = conversion;
  }

  /**
   * 从 Buff 数据创建 ConversionBuff 实例
   */
  static fromBuffData(data: any): ConversionBuff {
    // 解析 conversion
    if (!data.conversion) {
      console.error(`[DEBUG] ConversionBuff.fromBuffData: 缺少conversion字段`, data);
      throw new Error(`ConversionBuff.fromBuffData: 缺少conversion字段`);
    }
    
    const conversion = new Conversion(
      (PropertyType as any)[data.conversion.from_property],
      (PropertyType as any)[data.conversion.to_property],
      data.conversion.conversion_ratio,
      data.conversion.max_value
    );

    // 解析 target
    const targetData = data.target ?? {};
    const target = new BuffTarget();
    target.target_self = targetData.target_self ?? true;
    target.target_enemy = targetData.target_enemy ?? false;
    target.target_teammate = targetData.target_teammate ?? false;
    target.target_bund = targetData.target_bund ?? false;

    return new ConversionBuff(
      conversion,
      typeof data.source === 'string' ? (BuffSource as any)[data.source] : data.source,
      target,
      data.max_stacks ?? 1,
      data.stack_mode ?? 'linear',
      data.id ?? '',
      data.name ?? '',
      data.description ?? '',
      data.trigger_conditions ?? ''
    );
  }

  /**
   * 计算转换值
   */
  calculateConversion(sourceValue: number): number {
    let value = sourceValue * this.conversion.conversion_ratio;
    if (this.conversion.max_value !== undefined && value > this.conversion.max_value) {
      value = this.conversion.max_value;
    }
    return value;
  }
}
